### ISSUE TITLE
Suggest `;` on parse error when applicable

### ISSUE THREAD

**Issue by estebank** [rust-lang/rust#87197]
[Given the following code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5c83c78479ee9e0edd87b032e91d2200):

```rust
fn main() {
    let x = 100;
    println!("{}", x)
    let y = 200;
    println!("{}", y)
}
```

The current output is:

```
error: expected one of `.`, `;`, `?`, `}`, or an operator, found keyword `let`
 --> src/main.rs:4:5
  |
3 |     println!("{}", x)
  |                      - expected one of `.`, `;`, `?`, `}`, or an operator
4 |     let y = 200;
  |     ^^^ unexpected token
```

We should detect cases where a semicolon would be appropriate (probe that the current token and maybe limited lookahead confirm that what comes next could be a valid statement on its own) and suggest a semicolon. [If there is a missing `}`, we already suggest that it might belong there](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9e91b5712bd421147ab7fedaad331ebb). We might want to silence one or the other depending on whether the current token might be starting a statement or an item (`let` vs `struct`, etc.).

**Comment by amalik18** [comment]
Where would this stuff be contained?

**Comment by khang00** [comment]
Sorry, I don't have much time these days.

@rustbot release-assignment

**Comment by pierwill** [comment]
Just a guess: could this be addressed with another `else if self.look_ahead(1, |t|) {` block, as in https://github.com/rust-lang/rust/blob/3a24abd22fd25c836d8b4d75ff46c833f9c3934c/compiler/rustc_parse/src/parser/diagnostics.rs#L1168?

**Comment by pierwill** [comment]
I think we would need to answer this question: why is the logic here not handling the cases we want to handle?

https://github.com/pierwill/rust/blob/3a24abd22fd25c836d8b4d75ff46c833f9c3934c/compiler/rustc_parse/src/parser/diagnostics.rs#L1184-L1202

**Comment by pierwill** [comment]
> Just a guess: could this be addressed with another `else if self.look_ahead(1, |t|) {` block, as in
> 
> https://github.com/rust-lang/rust/blob/3a24abd22fd25c836d8b4d75ff46c833f9c3934c/compiler/rustc_parse/src/parser/diagnostics.rs#L1168
> 
> ?

This seems to be on track given the clue "probe that the current token and maybe limited lookahead confirm that what comes next could be a valid statement on its own"...

**Comment by estebank** [comment]
> I think we would need to answer this question: why is the logic here not handling the cases we want to handle?

That is a great question: as the code is written it looks like it _should_ work. I don't _know_ why it isn't triggering :(

**Comment by ebobrow** [comment]
I'd like to give this a try.

@rustbot claim
